testsuite/tests/memory-model/forbidden.ml

```diff
@@ -0,0 +1,748 @@
+        let old = try M.find k m with Not_found -> 0 in
+        M.add k (old+1) m
+
+      let pp chan m =
+        if Cfg.verbose then begin
+          Printf.fprintf chan "Test %s %s\n"
+            K.name (if K.allowed then "Allowed" else "Forbidden") ;
+          Printf.fprintf chan "Histogram (%d states)\n" (M.cardinal m) ;
+        end ;
+        let yes,no =
+          M.fold
+            (fun k n (yes,no) ->
+              let ok = K.ok k in
+              if Cfg.verbose then begin
+                let c = if ok then '*' else ':' in
+                Printf.fprintf chan "%8d%c> %a\n" n c K.pp k
+              end ;
+              if ok then (yes+n,no) else (yes,no+n))
+            m (0,0) in
+        let tag = match yes,no with
+          | 0,_ -> "Never"
+          | _,0 -> "Always"
+          | _,_ -> "Sometimes" in
+        if Cfg.verbose then begin
+          Printf.fprintf chan
+            "Observation %s %s %d %d\n" K.name tag yes no
+        end else begin
+          Printf.fprintf chan "Observation %s %s\n" K.name tag
+        end ;
+        if not K.allowed && yes <> 0 then begin
+          Printf.fprintf chan "Invalid behaviour on test %s\n" K.name
+        end ;
+        if Cfg.verbose then Printf.fprintf chan "\n%!"
+
+      let union = M.union (fun _ x y -> Some (x+y))
+    end
+end
+
+(****************)
+(* Run one test *)
+(****************)
+
+module type Test = sig
+  module Key : Hist.Key
+
+  module Env : sig
+    type t0
+    type t1
+    type t
+    val make : int -> t
+    val reinit : t -> unit
+    val env2in : t -> int -> (t0 * t1)
+  end
+
+  type out0
+  val no0 : out0
+  val code0 : (Env.t0 * Env.t1) -> out0
+
+  type out1
+  val no1 : out1
+  val code1 : (Env.t0 * Env.t1) -> out1
+
+  val out2key : (Env.t0 * Env.t1) -> out0 -> out1 -> Key.t
+end
+
+
+module MakeRun(C:Config)(T:Test) =
+  struct
+
+    module Hist = Hist.Make(C)(T.Key)
+
+    module Instance(_:sig end) = struct (* Applicative .. *)
+
+      let env = T.Env.make C.size
+      and barrier = Barrier.make C.size
+
+      let mkf id no code () =
+        let t = Array.make C.size no in
+        for i = 0 to C.size-1 do
+          Barrier.wait barrier id i ;
+          t.(i) <- code (T.Env.env2in env i)
+        done ;
+        t
+
+      let f0 = mkf 0 T.no0 T.code0
+      and f1 = mkf 1 T.no1 T.code1
+
+      let zyva () =
+        let hist = Hist.empty in
+        let rec run_rec hist j =
+          if j >= C.nruns then hist
+          else begin
+            T.Env.reinit env ;
+            Barrier.reinit barrier ;
+            let d0 = Domain.spawn f0 in
+            let d1 = Domain.spawn f1 in
+            let t0 = Domain.join d0 in
+            let t1 = Domain.join d1 in
+
+            let rec hist_rec hist i =
+              if i >= C.size then hist
+              else
+                let k = T.out2key (T.Env.env2in env i) t0.(i) t1.(i) in
+                hist_rec (Hist.see k hist) (i+1) in
+            let hist = hist_rec hist 0 in
```

Comment:
```ocaml
List.init C.size (fun i -> ...)
|> List.fold_left (fun h k -> Hist.see k h) hist
```

Reply:
I'd rather not build intermediate lists of length C.size or C.nruns. Although the suggested code is neat.